File: java/src/test/java/ai/vectros/smoke/CrossContext.java
package ai.vectros.smoke;

import ai.vectros.VectrosApiClient;
import ai.vectros.types.AccessProfileRequest;
import ai.vectros.types.AccessProfileRequestStatus;
import ai.vectros.types.AppContextRequest;
import ai.vectros.types.ScopeClause;
import ai.vectros.types.ScopedKeyResponse;
import ai.vectros.types.UserResponse;
import ai.vectros.resources.auth.requests.CreateAccessProfileRequest;
import ai.vectros.resources.auth.requests.CreateScopedKeyRequest;
import ai.vectros.resources.auth.requests.DeleteAppContextRequest;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ThreadLocalRandom;

/**
 * Two-app-context isolation fixture. An app context is a fail-closed data
 * partition within ONE tenant: a record/document
 * created while confined to context A must be invisible to a caller confined to
 * context B of the same tenant, on every read path. We provision two contexts,
 * each with a confined ssk_* client, and probe the partition over the real wire.
 */
final class CrossContext {
    private CrossContext() {}

    // Broad enough to create + read + delete within a context; the context
    // partition (not the scope) is what isolation under test relies on.
    static final List<String> ACTIONS = List.of(
        "records:c", "records:r", "records:d",
        "documents:c", "documents:r", "documents:d",
        "folders:c", "folders:r", "folders:d",
        "schemas:c", "schemas:r", "schemas:d",
        "search:r");

    static final class Handle {
        String contextId = "", userId = "", keyId = "";
        VectrosApiClient api;
    }

    static final class Fixture {
        final String tenantId;
        final Handle a, b;
        final List<Handle> created;
        Fixture(String tenantId, Handle a, Handle b, List<Handle> created) {
            this.tenantId = tenantId; this.a = a; this.b = b; this.created = created;
        }
        /** Best-effort teardown: revoke keys, cascade-delete contexts, delete users. */
        void teardown(VectrosApiClient root) {
            for (Handle h : created) {
                if (!h.keyId.isEmpty()) try { root.auth().revokeScopedKey(h.keyId); } catch (Exception ignore) {}
                if (!h.contextId.isEmpty()) try {
                    root.auth().deleteAppContext(h.contextId, DeleteAppContextRequest.builder().confirm(h.contextId).build());
                } catch (Exception ignore) {}
                if (!h.userId.isEmpty()) try { root.identity().deleteUser(h.userId); } catch (Exception ignore) {}
            }
        }
    }

    /** A context id valid under `^[a-z][a-z0-9-]{2,30}$`. */
    private static String contextId(String label) {
        String rnd = (System.currentTimeMillis() + Long.toString(Math.abs(ThreadLocalRandom.current().nextLong()), 36))
            .replaceAll("[^a-z0-9]", "");
        String id = "ctx-" + label + "-" + rnd;
        return id.length() > 31 ? id.substring(0, 31) : id;
    }

    private static Handle provision(VectrosApiClient root, String tenantId, String label, List<Handle> created) {
        Handle h = new Handle();
        h.api = root;
        created.add(h);

        h.contextId = contextId(label);
        root.auth().createAppContext(AppContextRequest.builder().contextId(h.contextId).name("cross-context " + label).build());

        UserResponse user = root.identity().createUser(ai.vectros.types.UserRequest.builder().externalId(Smoke.uniqueTag()).build());
        h.userId = user.getId().orElseThrow();

        root.auth().createAccessProfile(h.contextId, CreateAccessProfileRequest.builder().body(
            AccessProfileRequest.builder()
                .principalId("usr_" + h.userId)
                .scopes(List.of(ScopeClause.builder().allowedActions(ACTIONS).build()))
                .status(AccessProfileRequestStatus.ACTIVE)
                .build()).build());

        ScopedKeyResponse minted = root.auth().createScopedKey(CreateScopedKeyRequest.builder()
            .keyName("cross-context-" + label + "-" + Smoke.uniqueTag())
            .tenantId(tenantId).contextId(h.contextId).userId(h.userId).build());
        String raw = minted.getRawKey().orElseThrow(
            () -> new IllegalStateException("scoped key for context " + label + " returned no rawKey"));
        h.keyId = minted.getKeyId().orElse("");
        h.api = Smoke.client(raw);
        return h;
    }

    /** Two sibling contexts (A, B) in one tenant, each with a confined client. */
    static Fixture provisionTwo(VectrosApiClient root) {
        String tenantId = Smoke.env("VECTROS_LIVE_TENANT_ID");
        List<Handle> created = new ArrayList<>();
        try {
            Handle a = provision(root, tenantId, "a", created);
            Handle b = provision(root, tenantId, "b", created);
            return new Fixture(tenantId, a, b, created);
        } catch (RuntimeException e) {
            new Fixture(tenantId, null, null, created).teardown(root);
            throw e;
        }
    }
}

File: typescript/src/cross-context.ts
/**
 * Reusable two-app-context fixture for cross-CONTEXT data-isolation smokes.
 *
 * An app context is a fail-closed data partition within a single tenant: a
 * record/document/folder/schema created while operating in context A must be
 * invisible to a caller operating in context B of the SAME tenant, on every
 * read path. Cross-TENANT isolation is exercised elsewhere (a TEST-tenant key
 * cannot see LIVE-tenant content); this is the cross-CONTEXT analogue, the
 * higher-value seam because both contexts share one tenant — the only thing
 * keeping their data apart is the context partition itself.
 *
 * How a caller is CONFINED to a context over the wire
 * ----------------------------------------------------
 * A root server key operates cross-context (it can see every context's data),
 * so it cannot demonstrate confinement. A scoped server key (`ssk_*`) is bound
 * to an AccessProfile inside ONE context; the platform derives the caller's
 * context from that binding and applies it as a mandatory partition on every
 * finder. So to get a "context-B-confined caller" we:
 *
 *   1. create context B (root key),
 *   2. create a user + an AccessProfile for that user inside context B,
 *   3. mint an `ssk_*` key bound to that profile.
 *
 * The resulting key is a real, confined caller — exactly what you would
 * issue to a per-context workload. {@link provisionTwoContexts} builds two of
 * them (A and B) in one tenant and hands back a confined client for each.
 *
 * The fixture provisions only synthetic data and is self-cleaning. It is
 * deliberately self-contained so the resulting specs read as a standalone
 * proof of tenant/context isolation.
 */
import { VectrosClient } from '@vectros-ai/sdk';
import { client, getScopedClient } from './client';
import { uniqueTag, tryCleanup } from './helpers';

/**
 * The scope granted to each context's confined key. Broad enough to create and
 * read records/documents/folders/schemas and run search WITHIN the context;
 * the context partition — not the scope — is what isolation under test relies
 * on. No `dataScope` is declared, so the key sees all owners inside its own
 * context (and nothing outside it).
 */
const CONTEXT_KEY_SCOPES = [
    {
        allowed_actions: [
            'records:c', 'records:r', 'records:d',
            'documents:c', 'documents:r', 'documents:d',
            'folders:c', 'folders:r', 'folders:d',
            'schemas:c', 'schemas:r', 'schemas:d',
            'search:r',
        ],
    },
];

/** A single provisioned app context plus a client confined to it. */
export interface ContextHandle {
    /** The app-context id (the partition axis under test). */
    contextId: string;
    /** A real user in the tenant; the confined key binds to it. */
    userId: string;
    /** The minted `ssk_*` key's id — used to revoke it on teardown. */
    keyId: string;
    /** A VectrosClient whose every call is confined to {@link contextId}. */
    api: VectrosClient;
}

/** Two sibling contexts (A, B) in one tenant, each with a confined client. */
export interface TwoContextFixture {
    tenantId: string;
    ctxA: ContextHandle;
    ctxB: ContextHandle;
    /** Best-effort teardown: revoke keys, delete users, cascade-delete contexts. */
    teardown: () => Promise<void>;
}

/**
 * A context id valid under the platform's contextId rule
 * (`^[a-z][a-z0-9-]{2,30}$`): starts with a lowercase letter, lowercase
 * alphanumerics + hyphens, <= 31 chars. Purpose-built so it does not depend on
 * uniqueTag()'s incidental format staying within those bounds.
 */
function contextId(label: string): string {
    const rnd = `${Date.now()}${Math.random().toString(36).slice(2)}`.replace(/[^a-z0-9]/g, '');
    return `ctx-${label}-${rnd}`.slice(0, 31);
}

function requireTenantId(): string {
    const t = process.env.VECTROS_LIVE_TENANT_ID;
    if (!t) {
        throw new Error(
            'VECTROS_LIVE_TENANT_ID is required for cross-context isolation specs ' +
            '(the scoped-key mint endpoint needs the tenant the key will operate on). ' +
            'Set it in a .env file for local dev; in CI it is a pipeline variable.'
        );
    }
    return t;
}

/**
 * Tears down whatever a context handle created — revoke key, cascade-delete the
 * context (drains its records/documents/folders/schemas/profiles), delete the
 * (tenant-level) user. Best-effort: never throws.
 */
async function teardownHandle(h: ContextHandle): Promise<void> {
    if (h.keyId) {
        await tryCleanup(`revoke key ${h.keyId}`, () => client.auth.revokeScopedKey({ keyId: h.keyId }));
    }
    if (h.contextId) {
        // Confirm token must equal the contextId.
        await tryCleanup(`delete context ${h.contextId}`, () =>
            client.auth.deleteAppContext({ contextId: h.contextId, confirm: h.contextId }));
    }
    if (h.userId) {
        await tryCleanup(`delete user ${h.userId}`, () => client.identity.deleteUser({ id: h.userId }));
    }
}

/**
 * Provisions one context + a confined `ssk_*` client bound to it. Each created
 * resource is registered on `created` as it is created, so a failure partway
 * through can still be torn down by the caller (no orphaned context/user/key on
 * the shared tenant).
 */
async function provisionContext(tenantId: string, label: string, created: ContextHandle[]): Promise<ContextHandle> {
    const handle: ContextHandle = { contextId: '', userId: '', keyId: '', api: client };
    created.push(handle);

    handle.contextId = contextId(label);
    await client.auth.createAppContext({ body: { contextId: handle.contextId, name: `cross-context ${label}` } });

    // A real user is required: the key binds to an AccessProfile whose
    // principalId is `usr_<userId>` for this exact user.
    const user = await client.identity.createUser({ body: { externalId: uniqueTag() } });
    handle.userId = user.id!;

    await client.auth.createAccessProfile({
        contextId: handle.contextId,
        body: { principalId: `usr_${handle.userId}`, scopes: CONTEXT_KEY_SCOPES, status: 'active' },
    });

    const minted = await client.auth.createScopedKey({
        keyName: `cross-context-${label}-${uniqueTag()}`,
        tenantId,
        contextId: handle.contextId,
        userId: handle.userId,
    });
    if (!minted.rawKey) {
        throw new Error(
            `Scoped key for context ${label} returned no rawKey — cannot build a ` +
            `confined client. (rawKey is only returned once, on create.)`
        );
    }
    handle.keyId = minted.keyId!;
    handle.api = getScopedClient(minted.rawKey);
    return handle;
}

/**
 * Provisions TWO sibling contexts (A, B) in one tenant, each with its own
 * confined `ssk_*` client. Caller writes via one client and probes via the
 * other to prove the partition holds. Always pair with `teardown()` in an
 * afterAll. If provisioning fails partway, any already-created resources are
 * torn down before the error propagates — nothing orphans on the shared tenant.
 */
export async function provisionTwoContexts(): Promise<TwoContextFixture> {
    const tenantId = requireTenantId();
    const created: ContextHandle[] = [];
    let ctxA: ContextHandle;
    let ctxB: ContextHandle;
    try {
        ctxA = await provisionContext(tenantId, 'a', created);
        ctxB = await provisionContext(tenantId, 'b', created);
    } catch (err) {
        for (const h of created) await teardownHandle(h);
        throw err;
    }

    const teardown = async (): Promise<void> => {
        for (const h of created) await teardownHandle(h);
    };

    return { tenantId, ctxA, ctxB, teardown };
}
